text/0002-serializable-types.md

```diff
@@ -0,0 +1,413 @@
+
+MLAPI used to offer serialization support that is much less performant (due to boxing, runtime type checking etc.) and (arguably) less convenient. When [Standard RPC API](https://github.com/Unity-Technologies/com.unity.multiplayer.rfcs/blob/master/text/0001-std-rpc-api.md) was introduced, it also did open a whole set of opportunities in this area. At the time of writing this RFC, proposed design immediately applies to serialization of RPC parameters but it's very high likely that this design will also be used in other areas of type serialization such as network variables.
+
+We want to offer built-in serialization support for most commonly used types and also support custom serialization for user-defined types with ease.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Serialization
+
+Multiplayer framework has built-in serialization support for C# and Unity primitive types out-of-the-box, also with ability to further extend network serialization for user-defined types implementing `INetworkSerializable` interface.
+
+### C# Primitives
+
+`bool`, `char`, `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `float`, `double`, `string` types will be serialized by built-in serialization code.
+
+```cs
+[ServerRpc]
+void FooServerRpc(int somenumber, string sometext) { /* ... */ }
+
+void Update()
+{
+    if (Input.GetKeyDown(KeyCode.P))
+    {
+        FooServerRpc(Time.frameCount, "hello, world"); // Client -> Server
+    }
+}
+```
+
+### Unity Primitives
+
+`Color`, `Color32`, `Vector2`, `Vector3`, `Vector4`, `Quaternion`, `Ray`, `Ray2D` types will be serialized by built-in serialization code.
+
+```cs
+[ClientRpc]
+void BarClientRpc(Color somecolor) { /* ... */ }
+
+void Update()
+{
+    if (Input.GetKeyDown(KeyCode.P))
+    {
+        BarClientRpc(Color.red); // Server -> Client
+    }
+}
+```
+
+### Enum Types
+
+A user-defined enum type will be serialized by built-in serialization code (with underlying integer type).
+
+```cs
+enum SmallEnum : byte
+{
+    A,
+    B,
+    C
+}
+
+enum NormalEnum // default -> int
+{
+    X,
+    Y,
+    Z
+}
+
+[ServerRpc]
+void ConfigServerRpc(SmallEnum smallEnum, NormalEnum normalEnum) { /* ... */ }
+
+void Update()
+{
+    if (Input.GetKeyDown(KeyCode.P))
+    {
+        ConfigServerRpc(SmallEnum.A, NormalEnum.X); // Client -> Server
+    }
+}
+```
+
+### Static Arrays
+
+Static arrays like `int[]` will be serialized by built-in serialization code if their underlying type is either one of serialization supported types (e.g. `Vector3`) or if they implement `INetworkSerializable` interface.
+
+```cs
+[ServerRpc]
+void HelloServerRpc(int[] scores, Color[] colors) { /* ... */ }
+
+[ClientRpc]
+void WorldClientRpc(MyComplexType[] values) { /* ... */ }
+```
+
+### NetworkObject & NetworkBehaviour
```

Comment:
Are there ever circumstances where a network id could be passed via RPC or custom serialization before the network object has actually been spawned? This is less important for RPCs I think and more about how this serialization strategy applies to network vars / state. Because for example if one network behavior references another, depending on the order of spawning them the reference may not be able to be fulfilled because the object does not yet exist. Or, does the network spawning always happen before state deserialization? I'm not familiar with the internals but just curious if this type of ordering has been considered when it comes to object references

Reply:
I think MLAPI doesn't guarantee such consistent today but by sheer luck, it works fine most of the time (don't quote me for that). There is a bug SF guys raised internally a few days ago which I believe is related to spawning/referencing/replication issue you're talking about here. For now, it's unclear with no guarantees. Unless @TwoTenPvP knows better than me in this area.

Reply:
Today, at least for Network Vars, MLAPI checks if a networkId is valid to the recipient and if not, spawns a warning and, under circumstances involving message buffering, saves it up and tries again later (e.g. HandleNetworkedVarUpdate)

Similarly in BitReader.cs, when it's dealing with a game object (in ReadObjectPacked) it throws out a warning.

I don't think you'll every catch all the cases regardless of spawn order / deserialization because clients can wink out at any time.  So at first blush I would be skeptical about trying to make this water-tight / make it the serialization system's responsibility.

HOWEVER I would be very interested in exploring having a clean mechanism for user code to get a definable callback for when you de-serialize and find the NO doesn't exist on your side vs. the warning today.

Reply:
Yeah, I would be curious to see how circular references work for join-in-progress scenarios since one side of the reference would always be invalid until all objects are finished spawning. The buffering does sound like a pretty elegant solution for that case though. That said, this concern might be out of scope for this RFC